File: src/AdventOfCode2015/Maths.cs
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

namespace MartinCostello.AdventOfCode2015
{
    using System.Collections.Generic;
    using System.Linq;

    /// <summary>
    /// A class containing mathematics-related methods. This class cannot be inherited.
    /// </summary>
    internal static class Maths
    {
        /// <summary>
        /// Returns all the permutations of the specified collection of values.
        /// </summary>
        /// <typeparam name="T">The type of the values.</typeparam>
        /// <param name="collection">The collection to get the permutations of.</param>
        /// <returns>
        /// An <see cref="IEnumerable{T}"/> that returns the permutations of <paramref name="collection"/>.
        /// </returns>
        internal static IEnumerable<IEnumerable<T>> GetPermutations<T>(ICollection<T> collection)
        {
            return GetPermutations(collection, collection.Count);
        }

        /// <summary>
        /// Returns all the permutations of the specified collection of strings.
        /// </summary>
        /// <typeparam name="T">The type of the values.</typeparam>
        /// <param name="collection">The collection to get the permutations of.</param>
        /// <param name="count">The number of items to calculate the permutations from.</param>
        /// <returns>
        /// An <see cref="IEnumerable{T}"/> that returns the permutations of <paramref name="collection"/>.
        /// </returns>
        internal static IEnumerable<IEnumerable<T>> GetPermutations<T>(IEnumerable<T> collection, int count)
        {
            if (count == 1)
            {
                return collection.Select((p) => new[] { p });
            }

            return GetPermutations(collection, count - 1)
                .SelectMany((p) => collection.Where((r) => !p.Contains(r)), (set, value) => set.Concat(new[] { value }));
        }
    }
}

File: src/AdventOfCode.Site/scripts/ts/puzzles/Maths.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { from, IEnumerable } from 'linq-to-typescript';

export class Maths {
    static getPermutations<T>(collection: T[]): IEnumerable<IEnumerable<T>> {
        return Maths.getPermutationsWithCount(collection, collection.length);
    }

    static getPermutationsWithCount<T>(collection: T[], count: number): IEnumerable<IEnumerable<T>> {
        const enumerable = from(collection);
        if (count === 1) {
            return enumerable.select((p: T) => from([p]));
        }

        const permutations = Maths.getPermutationsWithCount(collection, count - 1);
        return Maths.selectMany(
            permutations,
            (p) => enumerable.except(p),
            (set, value) => Maths.append(set, value)
        );
    }

    private static append<T>(source: IEnumerable<T>, value: T): IEnumerable<T> {
        return source.concatenate(from([value]));
    }

    // Adapted from https://github.com/arogozine/LinqToTypeScript/blob/9f6fc3bac2f48f746257f9a41a4192c02b3aa733/src/sync/_private/selectMany.ts

    private static selectMany<TSource, TCollection, TResult>(
        source: IEnumerable<TSource>,
        collectionSelector: (x: TSource) => IEnumerable<TCollection>,
        resultSelector: (x: TSource, collection: TCollection) => IEnumerable<TResult>
    ) {
        function* iterator() {
            for (const element of source) {
                for (const subElement of collectionSelector(element)) {
                    yield resultSelector(element, subElement);
                }
            }
        }

        return from(new BasicEnumerable(iterator));
    }
}

class BasicEnumerable<TSource> implements Iterable<TSource> {
    public constructor(private readonly iterator: () => IterableIterator<TSource>) {}

    public [Symbol.iterator](): IterableIterator<TSource> {
        return this.iterator();
    }
}
